testing (dot) create dot unit tests

edwardmack opened the pull request:
## Changes

<!--

Please provide a brief but specific list of changes made, describe the changes
in functionality rather than the changes in code.

-->

- Create unit tests `dot` package: `build_spec_test.go`, `config_test.go`, `import_test.go`, `node_test.go`, `service_test.go` and `utils_test.go`
- 

## Tests

<!--

Details on how to run tests relevant to the changes within this pull request.

-->

```
go test ./dot/ -v

```

## Issues

<!--

Please link any issues that this pull request is related to and use the GitHub
supported format for automatically closing issues (ie, closes #123, fixes #123)

-->

- closes #1952 

## Primary Reviewer

<!--
Please indicate one of the code owners that are required to review prior to merging changes (e.g. @noot)
-->

- @timwu20, @qdm12

qdm12 commented:
I see a lot of integration tests deltas, is this expected? Did you maybe branch off your other integration tests PR?

> TestBuildSpec_ToJSON and TestBuildSpec_ToJSONRaw do not have test for errors with json.Marshal, I don't know how to create condition that will simulate an error with this.

JSON Marshal will fail if you try to marshal e.g. a function/channel. Maybe add some dummy function field to your `field` struct to trigger an error?

> TestWriteGenesisSpecFile doesn't test for errors from os.MkdirAll, I don't know how to create condition to simulate error.

Have a file at the same path 😉 Example

```
/tmp # touch test
/tmp # mkdir -p test
mkdir: can't create directory 'test': File exists
```

> since these functions build the struct I don't how to compare pointer addresses

Why would we care about the pointer address?

> Cleanup after errors

Shouldn't it cleanup even when there is no error? 🤔

### Other general comments

- For testing errors, I would suggest to use (since we already use testify around in the repo):

    ```go
    if testCase.err != nil {
    	assert.EqualError(t, err, testCase.err.Error())
    } else {
    	assert.NoError(t, err)
    }
    ```

- You can use `assert.Equal(t, expected, actual)` instead of `reflect.DeepEqual` + `t.Errorf(...)` since it does the same thing, it would accelerate test writing/shorten your test code 😉

edwardmack commented:
> I see a lot of integration tests deltas, is this expected? Did you maybe branch off your other integration tests PR?
> 
I did branch off my other integration tests PR, is this a bad idea?  Whats the best way to work on stuff that depends on other stuff you've done.  I've gotten into git branch trouble before, so I'm open to suggestions of best way to organize this.

> JSON Marshal will fail if you try to marshal e.g. a function/channel. Maybe add some dummy function field to your `field` struct to trigger an error?
> 
I'm no longer sure I need to test for this since the struct passed to this function insures it will not receive a function/channel.
 
> > TestWriteGenesisSpecFile doesn't test for errors from os.MkdirAll, I don't know how to create condition to simulate error.
> 
> Have a file at the same path wink Example
> 
> ```
> /tmp # touch test
> /tmp # mkdir -p test
> mkdir: can't create directory 'test': File exists
> ```

I wasn't able to create condition where `os.MkdirAll` failed.

> > since these functions build the struct I don't how to compare pointer addresses
> 
> Why would we care about the pointer address?
> 
Good point, I've updated tests to compare returned values.

> > Cleanup after errors
> 
> Shouldn't it cleanup even when there is no error? thinking
> 

Yes, good point, I've updated this.

> ### Other general comments
> * For testing errors, I would suggest to use (since we already use testify around in the repo):
>   ```go
>   if testCase.err != nil {
>   	assert.EqualError(t, err, testCase.err.Error())
>   } else {
>   	assert.NoError(t, err)
>   }
>   ```
> * You can use `assert.Equal(t, expected, actual)` instead of `reflect.DeepEqual` + `t.Errorf(...)` since it does the same thing, it would accelerate test writing/shorten your test code wink

Updated.

qdm12 commented:
> I did branch off my other integration tests PR, is this a bad idea? Whats the best way to work on stuff that depends on other stuff you've done. I've gotten into git branch trouble before, so I'm open to suggestions of best way to organize this.

I'd suggest you point this PR to that other branch you branched out from, merge the second one (this one) first and then the other branch (the first one) in development.

And yes don't worry too much about impossible to reproduce errors 😄 I'll mess around with `os.MkdirAll` though I'm curious now!

timwu20 commented:
Can you change the base branch of this PR to the other branch you moved the integration tests too? It'll make the diff easier to read.

timwu20 commented on dot/build_spec_test.go:
```diff
@@ -16,136 +16,193 @@
+	}{
+		{
+			name:   "normal conditions",
+			fields: fields{genesis: &genesis.Genesis{Name: "test"}},
```
can you add more tests to test the other attributes of `genesis.Genesis`

edwardmack replied:
added.

timwu20 commented:
I wouldn't worry about trying to test coverage up to 100%.  It's not really necessary to test `json.Marshal` and file read errors if you're already above 70% coverage.

qdm12 commented on dot/node.go:
```diff
@@ -180,7 +180,7 @@ func NodeInitialized(basepath string) bool {
 
 // LoadGlobalNodeName returns the stored global node name from database
-func LoadGlobalNodeName(basepath string) (nodename string, err error) {
+func LoadGlobalNodeName(basepath string) (string, error) {
```
Please let's use named returned values, it's really useful for readers/documentation.

edwardmack replied:
updated.

qdm12 commented on dot/services_test.go:
```diff
@@ -100,54 +214,124 @@ func Test_createCoreService(t *testing.T) {
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("createCoreService() got = %v, want %v", got, tt.want)
+			if tt.want != nil {
+				assert.NotNil(t, got)
```
Maybe assert it's nil if we expect it to be nil as well :thinking: ?

edwardmack replied:
Good idea, added.

qdm12 commented on dot/services_test.go:
```diff
@@ -100,54 +214,124 @@ func Test_createCoreService(t *testing.T) {
 
 func Test_createDigestHandler(t *testing.T) {
+	cfg := NewTestConfig(t)
+	require.NotNil(t, cfg)
```
nit I don't think this NotNil is really required since (afaik) we are asserting something from the test (`NewTestConfig`). Same for below checks as well. Although if these helper functions do use some production code, maybe do the assertion inside them since you are already passing `t` to them.

edwardmack replied:
Ok, removed.

qdm12 commented on dot/services_test.go:
```diff
@@ -247,22 +527,49 @@ func Test_createRuntime(t *testing.T) {
+				st:   stateSrvc,
+				code: []byte(`fake code`),
+			},
+			err: errors.New("failed to create runtime executor: Failed to instantiate the module:\n    compile error: Validation error \"Bad magic number\""),
```
Is it us or the runtime returning an error with `\n`?

qdm12 commented on lib/babe/verify.go:
```diff
@@ -58,12 +60,26 @@ type VerificationManager struct {
 func NewVerificationManager(blockState BlockState, epochState EpochState) (*VerificationManager, error) {
-	if blockState == nil {
-		return nil, errNilBlockState
+	switch blockState.(type) {
+	case *state.BlockState:
+		if blockState == (*state.BlockState)(nil) {
+			return nil, errNilBlockState
+		}
+	default:
+		if reflect.ValueOf(blockState).IsNil() {
+			return nil, errNilBlockState
+		}
 	}
```
supernit to reduce code duplication if we add more cases :wink: 

```suggestion
	var isNil bool
	switch blockState.(type) {
	case *state.BlockState:
		isNil = blockState == (*state.BlockState)(nil)
	default:
		isNil = reflect.ValueOf(blockState).IsNil()
	}
	if isNil {
		return nil, errNilBlockState
	}
```

edwardmack replied:
updated.

qdm12 commented on dot/node.go:
```diff
@@ -146,29 +146,27 @@ func NodeInitialized(basepath string) bool {
-		err = db.Close()
-		if err != nil {
-			logger.Errorf("failed to close database: %s", err)
+		e := db.Close()
```
nit, maybe instead of logging the close error:

1. Rename `e` to `closeErr` (sorry I'm more of an entire words person that single letter C style names 😄) 
2. If `err` (the returned error) is nil, set it to `closeErr`

edwardmack replied:
updated.

qdm12 commented on dot/node.go:
```diff
@@ -146,29 +146,27 @@ func NodeInitialized(basepath string) bool {
 	}
-
-	return nodename, err
+	return
```
No naked returns, it's really harder to read and leads to confusion

https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_naked_returns

To be fair, I have no clue why this is even part of the language...

edwardmack replied:
updated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ChainSafe/gossamer?pullRequest=1980) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ChainSafe/gossamer?pullRequest=1980) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: edwardmack<br/>:x: Ed Mack<br/><hr/>**Ed Mack** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ChainSafe/gossamer?pullRequest=1980) it.</sub>

qdm12 commented on dot/node.go:
```diff
@@ -326,6 +326,7 @@ func newNodeC(cfg *Config, nn newNodeIface) (*Node, error) {
 
 	return node, nil
 }
+//go:generate mockgen -source=node.go -destination=mock_node.go -package=dot
```
Change it to

```suggestion
//go:generate mockgen -destination=mock_node_test.go -package $GOPACKAGE . newNodeIface
```

So it's unexported to other packages (as a test file) and also using `source` isn't the best if it can be avoided. There are some corner cases where it's necessary (I don't remember which though haha), but it's better to just precise the interface name.

You should

edwardmack replied:
When I tried running this I get:
```
emack@system76-pc:~/projects/ChainSafe/gossamer/dot$ go generate
# command-line-arguments
./prog.go:27:38: cannot refer to unexported name dot.newNodeIface

```
so I needed to use the `-source` flag to solve issue.

qdm12 replied:
Totally! That was the case where you needed it 😄  Although you can use `$GOPACKAGE` instead of `dot` (if you want to)

qdm12 commented on dot/node_test.go:
```diff
@@ -200,17 +201,75 @@ func TestNewNodeC(t *testing.T) {
+			got, err := NewNodeC(tt.args.cfg)
+			if tt.err != nil {
+				assert.EqualError(t, err, tt.err.Error())
+				utils.RemoveTestDir(t)
```
nit maybe use `t.TestDir` instead?

edwardmack replied:
I'm not sure I understand what you mean here?

qdm12 replied:
I'm not sure where the directory is created but you can use `t.TestDir` to create a temp directory that gets removed at the end of the test. It might be out of scope here though.

qdm12 commented on dot/node_test.go:
```diff
@@ -200,17 +201,75 @@ func TestNewNodeC(t *testing.T) {
+				assert.NoError(t, err)
+			}
+
+			if tt.want != nil {
+				assert.Equal(t, tt.want.Name, got.Name)
+			}
```
Why not assert the entire Node instead of just its name?

edwardmack replied:
Updated.

qdm12 commented on dot/node_test.go:
```diff
@@ -200,17 +201,75 @@ func TestNewNodeC(t *testing.T) {
+	m.EXPECT().loadRuntime(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any())
+	m.EXPECT().createBlockVerifier(gomock.Any())
+	m.EXPECT().createDigestHandler(gomock.Any())
+	m.EXPECT().createCoreService(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any())
```
Please try to not use `gomock.Any()` since it covers code but test depth is quite low and some bugs can sneak in.

Ideally declare both the gomock controller and mocks inside the subtests with parameters from each test case to define the mock behavior.

edwardmack replied:
updated to remove use of `gomock.Any()`.

noot commented on dot/build_spec_test.go:
```diff
@@ -4,136 +4,345 @@
+			assert.Equal(t, tt.want, got)
+		})
+	}
+	//<<<<<<< HEAD
+	//=======
+	//
+	//	bs.genesis.ChainType = expectedChainType
+	//	bs.genesis.Properties = expectedProperties
+	//
+	//	require.NoError(t, err)
+	//
+	//	// confirm human-readable fields
```
remove commented out block

noot commented on dot/build_spec_test.go:
```diff
@@ -4,136 +4,345 @@
+	//		gen := new(genesis.Genesis)
+	//		err = json.Unmarshal(genesisBytes, gen)
+	//		require.NoError(t, err)
+	//>>>>>>> development
```
same here

noot commented on dot/node.go:
```diff
@@ -333,19 +386,52 @@ func NewNode(cfg *Config, ks *keystore.GlobalKeystore) (*Node, error) {
-	return node, nil
+}
+
+//InitKeystore to initialize node keystore
```
```suggestion
// InitKeystore initialised the node's keystore
```

noot commented on dot/node.go:
```diff
@@ -41,9 +48,87 @@ type Node struct {
 
+//go:generate mockgen -source=node.go -destination=mock_node_test.go -package=$GOPACKAGE
+
+type newNodeIface interface {
```
nice! I like this interface a lot, maybe name it `nodeBuilder` or `nodeCreator` or something like that?

noot commented on lib/babe/verify.go:
```diff
@@ -47,11 +49,25 @@ type VerificationManager struct {
 func NewVerificationManager(blockState BlockState, epochState EpochState) (*VerificationManager, error) {
-	if blockState == nil {
+	var isNilBlock bool
+	switch blockState.(type) {
+	case *state.BlockState:
+		isNilBlock = blockState == (*state.BlockState)(nil)
+	default:
+		isNilBlock = reflect.ValueOf(blockState).IsNil()
+	}
+	if isNilBlock {
 		return nil, errNilBlockState
 	}
 
-	if epochState == nil {
+	var isNilEpoch bool
+	switch epochState.(type) {
+	case *state.EpochState:
+		isNilEpoch = epochState == (*state.EpochState)(nil)
+	default:
+		isNilEpoch = reflect.ValueOf(epochState).IsNil()
+	}
+	if isNilEpoch {
```
do we ned this `isNilBlock = blockState == (*state.BlockState)(nil)` or does `reflect.ValueOf(blockState).IsNil()` evaluate to the same thing? I'd rather not have the `(*state.BlockState)` type assertion if possible

noot commented on lib/blocktree/blocktree.go:
```diff
@@ -403,7 +403,7 @@ func (bt *BlockTree) StoreRuntime(hash common.Hash, in runtime.Instance) {
 func (bt *BlockTree) GetBlockRuntime(hash common.Hash) (runtime.Instance, error) {
 	ins, ok := bt.runtime.Load(hash)
-	if !ok {
+	if !ok || ins == nil {
```
is `ins == nil` a case that happens? that seems concerning to me lol

noot reviewed (commented):
very nice work! just a few comments, but the tests look good to me :D

edwardmack commented:
closed, replaced by PR #2112
